src/crypto/muhash.cpp

```diff
@@ -0,0 +1,303 @@
+
+#include <cassert>
+#include <cstdio>
+#include <limits>
+
+namespace {
+
+/** 2^3072 - 1103717, the largest 3072-bit safe prime number, is used as the order of the group. */
+constexpr Num3072::limb_t MAX_PRIME_DIFF = 1103717;
+
+constexpr size_t INPUT_SIZE = 32;
+constexpr size_t OUTPUT_SIZE = 384;
+
+using limb_t = Num3072::limb_t;
+using double_limb_t = Num3072::double_limb_t;
+constexpr int LIMB_SIZE = Num3072::LIMB_SIZE;
+constexpr int LIMBS = Num3072::LIMBS;
+
+// Sanity check for Num3072 constants
+static_assert(LIMB_SIZE * LIMBS == 3072, "Num3072 isn't 3072 bits");
+static_assert(sizeof(double_limb_t) == sizeof(limb_t) * 2, "bad size for double_limb_t");
+static_assert(sizeof(limb_t) * 8 == LIMB_SIZE, "LIMB_SIZE is incorrect");
+
+// Hard coded values in MuHash3072 constructor and Finalize
+static_assert(sizeof(limb_t) == 4 || sizeof(limb_t) == 8, "bad size for limb_t");
+
+/** Extract the lowest limb of [c0,c1,c2] into n, and left shift the number by 1 limb. */
+inline void extract3(limb_t& c0, limb_t& c1, limb_t& c2, limb_t& n)
+{
+    n = c0;
+    c0 = c1;
+    c1 = c2;
+    c2 = 0;
+}
+
+/** Extract the lowest limb of [c0,c1] into n, and left shift the number by 1 limb. */
+inline void extract2(limb_t& c0, limb_t& c1, limb_t& n)
+{
+    n = c0;
+    c0 = c1;
+    c1 = 0;
+}
+
+/** [c0,c1] = a * b */
+inline void mul(limb_t& c0, limb_t& c1, const limb_t& a, const limb_t& b)
+{
+    double_limb_t t = (double_limb_t)a * b;
+    c1 = t >> LIMB_SIZE;
+    c0 = t;
+}
+
+/* [c0,c1,c2] += n * [d0,d1,d2]. c2 is 0 initially */
+inline void mulnadd3(limb_t& c0, limb_t& c1, limb_t& c2, limb_t& d0, limb_t& d1, limb_t& d2, const limb_t& n)
+{
+    double_limb_t t = (double_limb_t)d0 * n + c0;
+    c0 = t;
+    t >>= LIMB_SIZE;
+    t += (double_limb_t)d1 * n + c1;
+    c1 = t;
+    t >>= LIMB_SIZE;
+    c2 = t + d2 * n;
+}
+
+/* [c0,c1] *= n */
+inline void muln2(limb_t& c0, limb_t& c1, const limb_t& n)
+{
+    double_limb_t t = (double_limb_t)c0 * n;
+    c0 = t;
+    t >>= LIMB_SIZE;
+    t += (double_limb_t)c1 * n;
+    c1 = t;
+}
+
+/** [c0,c1,c2] += a * b */
+inline void muladd3(limb_t& c0, limb_t& c1, limb_t& c2, const limb_t& a, const limb_t& b)
+{
+    double_limb_t t = (double_limb_t)a * b;
+    limb_t th = t >> LIMB_SIZE;
+    limb_t tl = t;
+
+    c0 += tl;
+    th += (c0 < tl) ? 1 : 0;
+    c1 += th;
+    c2 += (c1 < th) ? 1 : 0;
+}
+
+/** [c0,c1,c2] += 2 * a * b */
+inline void muldbladd3(limb_t& c0, limb_t& c1, limb_t& c2, const limb_t& a, const limb_t& b)
+{
+    double_limb_t t = (double_limb_t)a * b;
+    limb_t th = t >> LIMB_SIZE;
+    limb_t tl = t;
+
+    c0 += tl;
+    limb_t tt = th + ((c0 < tl) ? 1 : 0);
+    c1 += tt;
+    c2 += (c1 < tt) ? 1 : 0;
+    c0 += tl;
+    th += (c0 < tl) ? 1 : 0;
+    c1 += th;
+    c2 += (c1 < th) ? 1 : 0;
+}
+
+/** [c0,c1] += a */
+inline void add2(limb_t& c0, limb_t& c1, const limb_t& a)
+{
+    c0 += a;
+    c1 += (c0 < a) ? 1 : 0;
+}
+
+bool IsOverflow(const Num3072& d)
```

Comment:
Hm, we use the same "overflow" terminology in secp256k1. The idea is that Num3072 represents a number in the range 0 to group order, so when it's larger it has overflown this range. (Idk, this may be a matter of taste...)

Reply:
Ah, I didn't realise that 'overflow' was the terminology used for "d is larger than the modulus". I suppose `IsNotLeastResidue()` would be most accurate, but as long as the terminology overflow is used consistently, I think just a comment is enough here.

Reply:
Added a comment for clarification.